### PR TITLE
WIP - context declarations

### PR DESCRIPTION
**Pull request by nemni8**
working! include manifest and pageelements map support.
tested in both flows, configPath and declarations.
Please ignore the dist and types folders, it will be removed from the final commit. i am using it to be able to consume this branch from code code before we published an official version.

### CODE REVIEW

**Comment by agankarin** [comment]
@nemni8 Can you please explain what should be the output after those changes?
Would `corvid-types` create a map for the dts files? would it bundle it to a different file?

About getting the path of both our declaration files and JS dts, while I was reading about the [language service API](https://github.com/microsoft/TypeScript/wiki/Using-the-Language-Service-API), I read that the compiler provides an API to get the source file list when you compile a program and this list already resolved the path we used whit the `reference path=..` tag.
I used your compiler function(in test) and the output was:
`/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es5.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2015.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2016.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2017.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2015.core.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2015.collection.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2015.generator.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2015.iterable.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2015.promise.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2015.proxy.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2015.reflect.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2015.symbol.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2015.symbol.wellknown.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2016.array.include.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2017.object.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2017.sharedmemory.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2017.string.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2017.intl.d.ts []
/users/karinag/projects/corvid-types/node_modules/typescript/lib/lib.es2017.typedarrays.d.ts []
/users/karinag/projects/corvid-types/types/common/emptybuffer.d.ts []
/users/karinag/projects/corvid-types/types/common/void$w$w.d.ts []
/users/karinag/projects/corvid-types/types/common/declaration.d.ts []
/users/karinag/projects/corvid-types/types/public/index.d.ts [ { pos: 21, end: 47, fileName: '../common/emptyBuffer.d.ts' },
  { pos: 73, end: 96, fileName: '../common/void$w$w.d.ts' },
  { pos: 122, end: 148, fileName: '../common/declaration.d.ts' } ]
`

File: scripts/generate-full-corvid-types.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env node
+const tmp = require("tmp");
+const fs = require("fs-extra");
+const path = require("path");
+const ts = require("typescript");
+const without_ = require("lodash/without");
+const TS_CONFIG_PATHS = {
```
Comment (dor-itzhaki) [inline scripts/generate-full-corvid-types.js]: unite with the other place where this is configured

File: scripts/generate-full-corvid-types.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+  const program = ts.createProgram({
+    rootNames: fileNames,
+    options
+  });
+
+  return program
+    .getSourceFiles()
+    .filter(file => file.isDeclarationFile)
+    .map(file => file.path);
```
Comment (dor-itzhaki) [inline scripts/generate-full-corvid-types.js]: convert to relative path instead of below

File: test/utils/tsconfig.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  "extends": "${path.relative(testPath, "configs/tsconfig.public.json")}",
+  "compilerOptions": {
+    "baseUrl": ".",
+    "paths": {
+      "public/*": ["./*"]
+    }
+  }
+}`;
+
+const getEmptyTsConfig = () => `{
```
Comment (dor-itzhaki) [inline test/utils/tsconfig.js]: expose this by itself, instead of getting it when you pass empty context

File: test/it/positive.spec.js [hunk 4]
```diff
@@ -1,13 +1,59 @@
...
+
 const POSITIVE_ROOT_PATH = "test/it/code-samples/positive";
 const positiveRoots = listSubDirectories(POSITIVE_ROOT_PATH);
 
-describe("typescript - positive scenarios", () => {
+describe("typescript - positive scenarios - configPaths flow", () => {
+  it.each(positiveRoots)(
+    "should successfully compile %s folder",
+    async tsRootPath => {
+      const testTmpDirPath = tmp.dirSync().name;
```
Comment (dor-itzhaki) [inline test/it/positive.spec.js]: try to put the "environment preparations" into a dedicated function, so it'll be easy to understand that we run the same tests with the 2 environments

File: src/widget.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        : `${value}(${getFunctionTypeParamsString(params)});`
+    }`;
+  }, "");
+};
+
+const getWidgetTypeDeclarations = ({ manifest }) => {
+  return Object.keys(manifest).reduce((dts, className) => {
+    const { members = [] } = manifest[className];
+    // need to understand which type should we extend, ($w.Node() or $w.Element)
+    return `${dts} declare namespace $w { interface ${className} extends $w.Node { ${getMembersTypes(
```
Comment (dor-itzhaki) [inline src/widget.js]: format this somehow?

File: test/it/code-samples/positive/pages-$w-dynamic-nicknames/widget.json [hunk 6]
```diff
@@ -0,0 +1,54 @@
+{
```
Comment (dor-itzhaki) [inline test/it/code-samples/positive/pages-$w-dynamic-nicknames/widget.json]: test with more than one widget

**Pull request merged**